src/Http/Http.Features/src/FeatureReferences.cs

```diff
@@ -86,15 +87,22 @@ public void Initalize(IFeatureCollection collection, int revision)
                 Cache = default;
             }
 
-            cached = Collection.Get<TFeature>();
+            var collection = Collection;
+            // Catch concurrent race conditions with Disposal
+            if (collection is null)
+            {
+                ThrowContextDisposed();
+            }
+
+            cached = collection.Get<TFeature>();
             if (cached == null)
             {
                 // Item not in collection, create it with factory
                 cached = factory(state);
                 // Add item to IFeatureCollection
-                Collection.Set(cached);
+                Collection?.Set(cached);
                 // Revision changed by .Set, update revision to new value
-                Revision = Collection.Revision;
+                Revision = Collection?.Revision ?? ContextDisposed();
```

Comment:
Why is this and the above line using and null-checking `Collection` instead of using `collection`?

Reply:
`collection` will be live (has passed null check) and not reflect whether the Context has been disposed; so will for example update Kestrel's feature collection for the next request rather than the "current" one.

Reply:
Its already passed a null check in `Fetch` before calling this so `Collection` gets nulled between the two calls

Reply:
Like you said in [your comment](https://github.com/dotnet/aspnetcore/issues/25454#issuecomment-686154020) on the issue, "this would mean the Context/Request/Response is disposed while the `Fetch` method is still accessed and the request is finished while it is still active, so concurrent activity; most likely from a missed `await`." That basically means all bets are off unless we start locking.

I don't want perfect to be the enemy of good. I totally support throwing an ODE instead of a NRE where we can, but this change allows `Fetch` to complete without throwing in situations it would throw and should throw today. We could be setting the `cached` item one `Collection` (or not set anything at all because it's null), but using the `Revision` of the next `Collection`. Today, if Collection is null when we try to set the cached item in the `Collection`, we'd throw an NRE instead of leaving `FeatureReferences` in an inconsistent state.

If we use `collection` instead of `Collection` here, at least the `FeatureReferences` stays in a consistent state. Sure, it could update the next request instead of the current one, but that's a real possibility with or without this change.

Reply:
>  this change allows `Fetch` to complete without throwing in situations it should throw and would throw today. 

It just changes it to throw ODE rather than NRE; using `collection` rather than `Collection` would all it to succeed as `FeatureReferences` is a struct which is "disposed" by being set to `default`
https://github.com/dotnet/aspnetcore/blob/e0467560e11a739897521326b543e9652a4edb77/src/Http/Http/src/Internal/DefaultHttpResponse.cs#L40-L43

Using `Collection` means it picks up that its been nulled; whereas using `collection` would mean it writes through ignoring that its been disposed and disconnected?

Reply:
If you are talking about skipping the `null` in the set and then passing as not `null` in the revision, that's a pretty small window; as essentially its only one null check immediately after another (since first would fail and not call `.Set`)
```csharp
Collection?.Set(cached);
Revision = Collection?.Revision ?? ContextDisposed();
```

Reply:
e.g. the current generate C# is
```csharp
IFeatureCollection collection = Collection;
if (collection == null)
{
    ThrowContextDisposed();
}
cached = collection.Get<TFeature>();
if (cached == null)
{
    cached = factory(state);
    IFeatureCollection collection2 = Collection;
    if (collection2 != null)
    {
        collection2.Set(cached);
    }
    IFeatureCollection collection3 = Collection;
    Revision = ((collection3 != null) ? collection3.Revision : ContextDisposed());
}
```

Reply:
> If you are talking about skipping the null in the set and then passing as not null in the revision, that's a pretty small window

That is what I'm talking about. Given there's already a null check in `Fetch`, this entire race is a pretty small window. That doesn't mean it won't happen.

Reply:
If it used `collection` then it wouldn't check for null again and everything would complete (albeit on the newer request); so it would reduce the chance of it throwing?

As it currently is it's more likely to throw a ODE than not throw? (in this narrow corner case)

Reply:
With this change, it's less likely to throw than before even when it should. By using `collection`, it would even be less likely to throw than with this change, but at least it wouldn't be getting a `Revision` from the new collection and allow the next call to `Fetch` to return a cached value that hasn't actually been set.

Reply:
> but at least it wouldn't be getting a `Revision` from the new collection and allow the next call to `Fetch` to return a cached value that hasn't actually been set.

On Kestrel the collection is the same; there isn't a new collection so it's going to return the same value for `Revision` regardless; so its better if it throws?

Reply:
Kestrel won't return the same value for Revision. 

https://github.com/dotnet/aspnetcore/blob/63116086f0098a7a8f2c5196d9fa778bc3df3b23/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.Generated.cs#L83-L88

My main objection isn't about any particular race condition being made better or worse though. I simply think this is too clever.

If we want to guard against concurrent access, we should lock, but I don't think it's worth it. I'm fine with null-checking at the beginning of Fetch like we already do, but repeatedly null-checking Collection is overly defensive in my opinion.